datalad_next/annexremote/uncurl.py

```diff
@@ -0,0 +1,470 @@
+"""
+*uncurl* git-annex external special remote
+==========================================
+
+This implementation is a git-annex accessible interface to datalad-next's URL
+operations framework. It serves two main purposes:
+
+1. Combine git-annex's capabilities of registering and accessing file content
+   via URLs with DataLad's access credential management and (additional or
+   alternative) transport protocol implementations.
+
+2. Minimize the maintenance effort for datasets (primarily) composed from
+   content that is remotely accessible via URLs from systems other than
+   Datalad or git-annex in the event of an infrastructure transition
+   (e.g. moving to a different technical system or a different data
+   organization on a storage system).
+
+
+Download helper with credential management support
+--------------------------------------------------
+
+The simplest way to use this remote is to initialize it without any particular
+configuration::
+
+    $ git annex initremote uncurl type=external externaltype=uncurl encryption=none
+    initremote uncurl ok
+    (recording state in git...)
+
+Once initialized, or later enabled in a clone, ``git-annex addurl`` will check
+with the *uncurl* remote whether it can handle a particular URL, and will let
+the remote perform the download in case of positive response. By default, the
+remote will claim any URLs with a scheme that the local datalad-next
+installation supports. This always includes ``file://``, ``http://``, and
+``https://``, but is extensible, and a particular installation may also support
+``ssh://`` (by default when openssh is installed), or other schemes.
+
+With this setup, download requests now use DataLad's credential system for
+authentication. DataLad will automatically lookup matching credentials, prompt
+for manual entry if none are found, and offer to store them securely for later
+use after having used them successfully::
+
+    $ git annex addurl http://httpbin.org/basic-auth/myuser/mypassword
+    Credential needed for access to http://httpbin.org/basic-auth/myuser/mypassword
+    user: myuser
+    password: 
+    password (repeat): 
+    Enter a name to save the credential
+    (sniffing http://httpbin.org/basic-auth/myuser/mypassword) securely for future
+    re-use, or 'skip' to not save the credential
+    name: httpbin-dummy
+
+    addurl http://httpbin.org/basic-auth/myuser/mypassword (from uncurl) (to ...) 
+    ok
+    (recording state in git...)
+
+By adding files via downloads from URLs in this fashion, datasets can be built
+that track information across a range of locations/services, using a possibly
+heterogeneous set of access methods.
```

Comment:
overall it feels like a replacement/next generation of the `datalad` annex special remote in datalad "core" since it works exactly as `datalad` one just uses the credentials/"downloader" of -next instead of core. I think it might be worth making it explicit.

Reply:
Thanks for the feedback. If that is true, it likely matches the behavior without url templates and match expressions configured.

Declaring feature parity would then depend on the capabilities of the `UrlOperations` in -next vs the downloaders in core. I think there is a substantial overlap, but I am not confident that it is "same". So such a declaration would need to be made with care.

That being said, feature parity is certainly a goal here (within reason).

Reply:
I added a statement to the documentation.